### PR TITLE
feat(eagle3): add Llama EAGLE-3 draft training recipe

### PR DESCRIPTION
**Pull request by khazic**
# What does this PR do?

Adds a minimal Llama-only EAGLE-3 speculative-decoding draft-model training recipe. The recipe trains a small draft head against a frozen Llama target's auxiliary hidden states and projected logit distribution, using the standard EAGLE-3 test-time-training (TTT) unroll.

# Changelog

**New module under `nemo_automodel/components/speculative/eagle/`:**

- `draft_llama.py` — `LlamaEagle3DraftModel` (single-layer decoder consuming a `[input_emb, projected_hidden]` 2H feature) plus its `Eagle3LlamaAttention`, `Eagle3LlamaMLP`, `Eagle3LlamaDecoderLayer` building blocks.
- `target.py` — `HFEagle3TargetModel` wrapper that registers forward hooks on a frozen target `LlamaForCausalLM` to capture three auxiliary hidden states (default layers `[1, n/2-1, n-4]`) and shift target logits / input_ids by one position.
- `core.py` — `Eagle3TrainerModule` runs the multi-step TTT unroll, shifting `input_ids` / `loss_mask` / `position_mask` / `target_probs` in lockstep, and aggregates per-step soft CE losses plus a top-1 accuracy metric.
- `loss.py` — `masked_soft_cross_entropy`, a KL-style soft CE normalized by the number of valid supervised positions.
- `data.py` — `build_eagle3_dataloader` plus `build_eagle3_token_mapping` (DDP-consistent draft-vocab construction via a dense `all_reduce`-sum of supervised-token frequencies and a stable `argsort`).

**New recipe entry point `nemo_automodel/recipes/llm/train_eagle3.py` (`TrainEagle3Recipe`):**

- Builds the draft model in the target's compute dtype (bf16 on cuda, fp32 on cpu) so linears, embeddings, and RMSNorm parameters stay dtype-consistent.
- Filters frozen parameters out of the optimizer.
- Applies a warmup + cosine LR schedule via two optimizer knobs: `warmup_ratio` and `min_lr_ratio`.
- Logs `Training start`, per-epoch end, per-checkpoint paths, and `Training complete` so a run is never silently exhausted.

**Examples and tests:**

- `examples/eagle3/llama_eagle3_mvp.yaml` — minimal smoke config (Llama-3.2-1B, placeholder local data path).
- `examples/eagle3/llama_eagle3_perfectblend.yaml` — production-scale config against `frankleeeee/PerfectBlend-Regenerated-Llama-3.1-8B-Instruct` with a `train[:200000]` slice (~6.25k optim steps, ~2h/epoch on 8 GPUs).
- `tests/unit_tests/speculative/test_eagle3.py` — unit tests covering the masked soft CE, target-distribution projection, multi-step TTT forward/backward (regression guard for a slice-vs-shift bug caught during development), draft forward shape, token-mapping frequency selection, and the shift helper.

# Design Rationale

This PR intentionally targets the simplest correct MVP rather than a fully-featured implementation. To validate the core algorithmic correctness of the EAGLE-3 training loop within this framework:

- **Model scope:** Implementation starts with Llama as the target architecture, since it is well-understood and widely used. This provides a stable, reproducible baseline for verifying correctness before extending to other model families.
- **Dataset:** `frankleeeee/PerfectBlend-Regenerated-Llama-3.1-8B-Instruct` was selected as it is a proven, publicly available dataset that has been validated for speculative-decoding draft training. A full 1M-sample run is currently in progress locally; results will be shared once complete.

Follow-up PRs will incrementally extend coverage:

- Support for additional target model families beyond Llama.
- Flash Attention / SDPA integration in the draft model for throughput.
- KV-cache recurrence across TTT steps for full algorithmic parity with the SpecForge reference.
- Resume support including optimizer and LR scheduler state.

> **Note:** This is the minimal correct version of the recipe, not the final form. Development will continue through subsequent PRs.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?

# Additional Information

- Related to # (issue)

This is the first speculative-decoding recipe under `nemo_automodel/components/speculative/`. Possible follow-ups:

- KV-cache recurrence across TTT steps, for full algorithmic parity with the SpecForge reference.
- SDPA-based attention in the draft for throughput.
- Resume support that includes optimizer / LR scheduler state.

### CODE REVIEW

File: examples/speculative/eagle3/llama_eagle3_mvp.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+recipe: TrainEagle3Recipe
+
+dist_env:
+  backend: nccl
+  timeout_minutes: 30
+
+recipe_args:
+  target_model_name_or_path: meta-llama/Llama-3.2-1B
+  train_data_path: /path/to/train.jsonl
```
Comment (HuiyingLi) [inline examples/speculative/eagle3/llama_eagle3_mvp.yaml:24]: Is there example what the jsonl looks like? Maybe we can do a follow up PR for the docs.

**Comment by khazic** [comment]
## CI status

The two L0_Unit_Tests failures
(``test_vlm_kd_uses_tp_kd_loss_path`` and
``test_vlm_kd_cp_prepare_feeds_student_inputs_embeds_to_cp_and_teacher``)
are unrelated to this PR — they were already red on ``main`` after #2211
+ #2215 landed independently and crossed a missing fixture attribute.
A one-line fix is up in #2225. Once #2225 merges, please re-run CI on
this PR (or rebase).

## End-to-end training run

Single-epoch training run on the example
``examples/eagle3/llama_eagle3_perfectblend.yaml`` config
(Llama-3.1-8B target, 200k PerfectBlend slice, 8 GPUs, peak LR 2e-4,
``ttt_steps=4``):

```
step=20    train_loss=10.37  train_acc=0.0000   (random init, ~ ln(32000))
step=2000  train_loss=~3.5   train_acc=~0.55
step=5400  train_loss=1.4406 train_acc=0.7211
step=6000  train_loss=1.3820 train_acc=0.7315
step=6240  train_loss=1.3856 train_acc=0.7317  lr=2.000e-05  (cosine min)
Epoch 0 done: total_batches_seen=25000 global_step=6250
```

End-of-epoch draft accuracy is **~73%** across the
``Σ_{i=0..3} 0.8^i * step_loss_i / Σ_{i=0..3} 0.8^i`` weighted-mean
TTT loss. LR followed the warmup + cosine schedule from peak ``2e-4``
to ``min_lr_ratio=0.1`` (= ``2e-5``) over the full epoch without
spikes. Gradient clip (``max_grad_norm=1.0``) was never load-bearing.

This confirms the end-to-end recipe (KV cache + diagonal attention,
position shift, decay-weighted mean loss, draft vocab mapping)
converges cleanly on a realistic dataset.

File: examples/speculative/eagle3/llama_eagle3_mvp.yaml
Comment (HuiyingLi) [inline examples/speculative/eagle3/llama_eagle3_mvp.yaml:1]: What about examples/speculative as a top level folder

**Comment by khazic** [comment]
Addressed the review feedback in ``302eb23a refactor(eagle3): address PR #2224 review feedback``:

- **Maintainer (HuiyingLi)**
  - ``examples/eagle3/`` → ``examples/speculative/eagle3/``. Good call -- gives Medusa / other speculative recipes a home next door without us reshuffling later.
  - jsonl format docs: agreed, will spin a follow-up docs PR once this lands so it can reference the merged paths.

- **claude bot**
  - ``target.py:_default_aux_layer_ids``: applied the order-preserving dedup suggestion. ``num_layers=5`` now returns ``[1]`` and ``num_layers=6`` returns ``[1, 2]`` instead of colliding hooks.
  - Missing ``HFEagle3TargetModel`` coverage: added ``tests/unit_tests/speculative/test_eagle3_target.py`` (5 new tests) covering the dedup, hook-capture concat-by-layer (cross-validated against HF's ``output_hidden_states=True``), the left-shift transformation on ``input_ids`` / ``loss_mask`` / ``logits``, the non-shifted ``attention_mask`` invariant, and the out-of-bounds ``aux_layer_id`` guard. 19/19 speculative unit tests passing locally.

File: examples/speculative/eagle3/llama_eagle3_mvp.yaml (hunk 1 above)
Comment (khazic) [reply examples/speculative/eagle3/llama_eagle3_mvp.yaml:24]: Added an inline OpenAI-chat schema example next to ``train_data_path`` in ``llama_eagle3_mvp.yaml`` (b6ac5269). Hopefully that's enough self-documentation for the recipe to be picked up without grepping into ``ChatDataset``; happy to still spin a follow-up docs PR once this lands if you'd like a longer-form ``examples/speculative/README.md``.

File: nemo_automodel/components/datasets/llm/eagle3.py
Comment (HuiyingLi) [inline nemo_automodel/components/datasets/llm/eagle3.py:1]: What about components/datasets/

File: nemo_automodel/components/loss/soft_ce.py
Comment (HuiyingLi) [inline nemo_automodel/components/loss/soft_ce.py:1]: What about components/loss

File: nemo_automodel/components/datasets/llm/eagle3.py
Comment (khazic) [reply nemo_automodel/components/datasets/llm/eagle3.py:1]: Moved to components/datasets/llm/eagle3.py in ed063000. build_eagle3_dataloader is a thin ChatDataset wrapper and build_eagle3_token_mapping builds the draft sub-vocab -- both fit the datasets/llm namespace alongside the other LLM dataset helpers. Imports in train_eagle3.py / core.py / tests now pull from the new path directly.

File: nemo_automodel/components/loss/soft_ce.py
Comment (khazic) [reply nemo_automodel/components/loss/soft_ce.py:1]: Moved to components/loss/soft_ce.py in ed063000. masked_soft_cross_entropy is a generic soft-target CE that any future distillation recipe could reuse; nothing about it is EAGLE-3-specific, so it belongs in the cross-cutting components/loss/ namespace next to masked_ce.py / kd_loss.py.

**Comment by hemildesai** [comment]
Can we add a toy L2 functional test for spec decoding in CI?

File: nemo_automodel/components/speculative/eagle/draft_llama.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        super().__init__(config)
+        self.config = config
+        self.target_hidden_size = getattr(config, "target_hidden_size", config.hidden_size)
+        self.draft_vocab_size = getattr(config, "draft_vocab_size", config.vocab_size)
+
+        self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size, padding_idx=config.pad_token_id)
+        self.hidden_proj = nn.Linear(self.target_hidden_size * 3, config.hidden_size, bias=False)
+        self.decoder = Eagle3LlamaDecoderLayer(config)
+        self.norm = initialize_rms_norm_module("torch", config.hidden_size, eps=config.rms_norm_eps, device=None)
+        self.lm_head = nn.Linear(config.hidden_size, self.draft_vocab_size, bias=False)
```
Comment (HuiyingLi) [inline nemo_automodel/components/speculative/eagle/draft_llama.py:258]: ~~The draft model has its own lm head instead of reusing target's?~~ eagle3 has it

**Review by HuiyingLi — APPROVED** [review]
Thank you so much!

**Comment by HuiyingLi** [comment]
> Can we add a toy L2 functional test for spec decoding in CI?

Thank you @hemildesai  will add L2 functional test in a followup PR.

**Comment by HuiyingLi** [comment]
@hemildesai  The L2 tests are added in #2254
